File: leetcode-js/092-reverse-linked-list-ii.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} m
 * @param {number} n
 * @return {ListNode}
 */
var reverseBetween = function(head, m, n) {
    if (!head) {
        return null;
    }

    const dummy = new ListNode(0);
    dummy.next = head;
    let prev = dummy;

    for (let i = 1; i < m; i += 1) {
        prev = prev.next;
    }

    let start = prev.next;
    let next = start.next;

    for (let i = 1; i <= n - m; i += 1) {
        start.next = next.next;
        next.next = prev.next;
        prev.next = next;
        next = start.next;
    }

    return dummy.next;
};

File: leetcode-java/092-reverse-linked-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;

        ListNode pre = dummy;

        for (int i = 1; i < m; i++) {
            pre = pre.next;
        }

        ListNode start = pre.next;
        ListNode cur = pre.next;

        for (int i = m; i < n; i++) {
            pre.next = cur.next;
            cur.next = pre.next.next;
            pre.next.next = start;

            start = pre.next;
        }

        return dummy.next;
    }
}
